doc/govuk-components.md

```diff
@@ -9,10 +9,58 @@ The available components and their documentation are exposed by an API at `/temp
 
 * a [Partial View](../app/views/govuk_component) - The template logic and markup, also defines the arguments expected
 * a [SCSS module](../app/assets/stylesheets/govuk-component) - The styling of the component
-* a Javascript module - no examples yet.
+* a [Javascript module](../app/assets/javascripts/govuk-component) - JS behaviour of the component
 * [Documentation](../app/views/govuk_component/docs) - a `.yml` per component, describing the component and containing fixture data.
 * a [unit test](../test/govuk_component) - testing the component renders correctly based on parameters passed in
 
+## Principles behind components
+
+### Encapsulate common UI
+
+GOV.UK Components aim to hide the complexity of a piece of common UI. Putting the templating logic, HTML, CSS and JS behind an abstraction, so that an app rendering a component only needs to know about the parameters to render it with. This is similar to how [React components](https://facebook.github.io/react/index.html) work.
+
+Think of it as a function call in a programming language, as long as the [interface](https://en.wikipedia.org/wiki/Application_programming_interface) doesn't change, then code rendering a component doesn't need to care about the changes.
+
+Here's an example of calling a component. The parameter is `title` with a value of `"My page title"`.
+
+```ruby
+<%= render partial: 'govuk_component/title', locals: { title: "My page title" } %>
+```
+
+This will output something like:
+
+```html
+<div class="govuk-title length-average">
+  <h1>My page title</h1>
+</div>
+```
+
+Apps should treat the output of a component like a [black box](https://en.wikipedia.org/wiki/Black_box), not relying on the internal implementation as it may change (eg, additional CSS based on the `govuk-title` or `h1` selector)
+
+If additional behavior in the component is needed then it should be made in the component itself, then used in an application.
```

Comment:
I think [higher-order functions/components](https://facebook.github.io/react/blog/2016/07/13/mixins-considered-harmful.html#higher-order-components-explained) (functions that augment other functions and return a new, better one) is also a good pattern to re-use certain types of behaviours between multiple components.